### PR TITLE
wayland/compositor: add wl_compositor@v6 constructor

### PR DESCRIPTION
**Pull request by kchibisov**
`wl_compositor@v6` requires special handling in downstream compositors, so smithay can't provide it unconditionally, thus allow setting a version for `wl_compositor` when creating a global.

--

This is probably not really a good solution, Alternative could be to have a separate constructor which builds _at least_ v6.

### CODE REVIEW

**Comment by kchibisov** [comment]
I'll fix CI after ack on approach, since change in it is a rewrite of the whole thing.

**Comment by YaLTeR** [comment]
Hm, which special handling is *required*?

**Comment by kchibisov** [comment]
>Hm, which special handling is required?

Downstream has to actually send the events for all the surfaces, it's not as automated as with `wl_output` and downstream could decide to use different scale factors for multiple windows on the same output.

**Comment by YaLTeR** [comment]
> Downstream has to actually send the events for all the surfaces

Do you mean preferred_buffer_scale/transform? I couldn't find where it says that the compositor should send them?

**Comment by YaLTeR** [comment]
Hm, I don't see where it says that in the protocol? I'd assume that if a surface never gets a preferred buffer scale, it should keep using wl_output scale?

**Comment by kchibisov** [comment]
Should be ready to go.

File: src/wayland/compositor/mod.rs [hunk 1]
```diff
@@ -564,18 +564,43 @@ impl CompositorClientState {
...
+    /// The same as [`new`], but binds at least version 6 of [`wl_compositor`].
+    ///
+    /// This means that for clients to scale and apply transformation with
+    /// non-default values [`wl_surface::preferred_buffer_scale`] and
+    /// [`wl_surface::preferred_buffer_transform`] must be sent.
+    ///
+    /// [`new`]: Self::new
+    /// [`wl_compositor`]: wayland_server::protocol::wl_compositor
+    /// [`wl_surface::preferred_buffer_scale`]: WlSurface::preferred_buffer_scale
+    /// [`wl_surface::preferred_buffer_transform`]: WlSurface::preferred_buffer_transform
```
Comment (Drakulix) [inline src/wayland/compositor/mod.rs:591]: No sorry, we are not asking downstream to send raw events. Downstream would have to check the version of global the client has bound to do this correctly, etc. This needs a function similar to `set_preferred_scale` for the fractional scale protocol.

**Comment by kchibisov** [comment]
I'm not sure I like the way it's because downstream can't call `send_surface_state` during the `with_surfaces`, since the data can only be mutated. Maybe this cache should all go into the `SurfaceData` instead of `PrivateSurfaceData` but that would mean that you need to get one cache first and then access the other one with the interior mutability?

The same issue is if you want to call basically any `compositor/mod.rs` api during the `with_surfaces`, but I guess you can avoid doing so for some, it's just most of them are just getters, but in this particular case you'd need an API which can actually set things...

**Comment by kchibisov** [comment]
A particular place I don't like with the given approach is https://github.com/kchibisov/niri/blob/37692b264f93e6018d0af04d7459b538e51618ae/src/layout.rs#L1958 , but it's not like we want the method on the `SurfaceData` or require `SurfaceData` before calling this function, since it could be annoying for downstream.

**Comment by kchibisov** [comment]
Added a separate commit for `new_subsuface`, so downstream can send a scale right away when parent is known and they can't present parent across the outputs.

**Comment by Drakulix** [comment]
> A particular place I don't like with the given approach is https://github.com/kchibisov/niri/blob/37692b264f93e6018d0af04d7459b538e51618ae/src/layout.rs#L1958 , but it's not like we want the method on the `SurfaceData` or require `SurfaceData` before calling this function, since it could be annoying for downstream.

Honestly I don't know. We require the `SurfaceData` in other places and it has been fine for downstream?
I don't have a particular strong preference here, but taking `SurfaceData` instead of the surface for `send_surface_state` would be more consistent with other apis.

>  Added a separate commit for new_subsuface, so downstream can send a scale right away when parent is known and they can't present parent across the outputs.

That seems like a good addition. :+1: 

Little bike-shed: I am not a huge fan of how we are overloading the word `state` through the whole crate.
Any other ideas for `send_surface_state`? It's a little long, but maybe even something like `send_surface_scale_transform` or `send_preferred_scale_transform`? (Or split the two up? Though that might cause some unnecessary overhead..)

**Comment by kchibisov** [comment]
>Honestly I don't know. We require the SurfaceData in other places and it has been fine for downstream?
I don't have a particular strong preference here, but taking SurfaceData instead of the surface for send_surface_state would be more consistent with other apis.

Yeah, but it's just other APIs don't mutate... I'll see how it looks though.

>send_preferred_scale_transform

Ok, but once something new is added it'll go `scale_transform_subpixel`? The state is inline with what it actually is, since `xdg-shell` also uses the combination `surface state` https://gitlab.freedesktop.org/wayland/wayland-protocols/-/blob/main/stable/xdg-shell/xdg-shell.xml?ref_type=heads#L439 .

**Comment by Drakulix** [comment]
> Ok, but once something new is added it'll go `scale_transform_subpixel`? The state is inline with what it actually is, since `xdg-shell` also uses the combination `surface state` https://gitlab.freedesktop.org/wayland/wayland-protocols/-/blob/main/stable/xdg-shell/xdg-shell.xml?ref_type=heads#L439 .

That is somewhat assuring. I guess as long as we don't have a better idea, lets keep it like this for now.

**Review by Drakulix — APPROVED** [review]
LGTM.

**Pull request merged**